## Set up

In [1]:
import os
import re
import string
import pandas as pd
import numpy as np

In [2]:
%load_ext autoreload

In [3]:
%reload_ext autoreload

In [4]:
os.getcwd()

'/Users/alessiatosi/DS_projects/behavioural-sci-perception/notebooks'

In [5]:
pd.options.display.max_seq_items = 10000
pd.set_option('display.max_colwidth', None)

Load environment variables

In [6]:
DATA_PATH = os.environ.get("DIR_DATA_RAW")

In [7]:
DATA_PATH

'/Users/alessiatosi/DS_projects/behavioural-sci-perception/data/raw'

In [8]:
OUTPUT_DIR = os.environ.get("DIR_DATA_INTERIM")

Constants

In [9]:
TWEETS_FLNM = "tweet-ids-British-Oct2020" 

In [10]:
VARS_TO_KEEP = ['created_at', 'hashtags',
       'favorite_count', 'id', 'reweet_id', 'retweet_screen_name', 'lang', 'place',
       'possibly_sensitive', 'retweet_count', 'text',
       'user_screen_name', 'user_followers_count', 'user_location', 'user_name', 'user_screen_name.1',
       'user_time_zone']

## Get data

In [11]:
tweets_raw = pd.read_csv(os.path.join(DATA_PATH, TWEETS_FLNM + '.csv'))

In [12]:
tweets_raw.shape

# Missing tweets: were they not hydrated because the tweet was not found or because of rate limit of the Twitter API?

(12161, 34)

In [13]:
tweets_raw.columns

Index(['coordinates', 'created_at', 'hashtags', 'media', 'urls',
       'favorite_count', 'id', 'in_reply_to_screen_name',
       'in_reply_to_status_id', 'in_reply_to_user_id', 'lang', 'place',
       'possibly_sensitive', 'retweet_count', 'reweet_id',
       'retweet_screen_name', 'source', 'text', 'tweet_url', 'user_created_at',
       'user_screen_name', 'user_default_profile_image', 'user_description',
       'user_favourites_count', 'user_followers_count', 'user_friends_count',
       'user_listed_count', 'user_location', 'user_name', 'user_screen_name.1',
       'user_statuses_count', 'user_time_zone', 'user_urls', 'user_verified'],
      dtype='object')

In [14]:
# filter only relevant columns
tweets_raw = tweets_raw[VARS_TO_KEEP]

In [15]:
tweets_raw[:3]

created_at             hashtags  favorite_count  \
0  Sun Apr 19 11:29:34 +0000 2020                  NaN               0   
1  Wed Mar 04 04:35:46 +0000 2020                  NaN               0   
2  Wed Apr 08 17:21:54 +0000 2020  COVID19 coronavirus               0   

                    id     reweet_id retweet_screen_name lang place  \
0  1251835324801613825  1.251829e+18      Jeremy11223344   en   NaN   
1  1235061349685719047  1.234871e+18          bmj_latest   en   NaN   
2  1247937727699484673  1.247901e+18         jayvanbavel   en   NaN   

  possibly_sensitive  retweet_count  \
0                NaN              3   
1                NaN            194   
2                NaN             89   

                                                                                                                                               text  \
0      RT @Jeremy11223344: @tjc1_tom @THemingford @dsue1441_re Imperial College modellers in collaboration with the Downing St "nudge" unit cannot…   
1      RT @bmj_latest: Human behaviour will determine how quickly covid-19 spreads, say @SusanMichie @DrRichardAmlot and @robertjwest What needs t…   
2  RT @jayvanbavel: We posted a revised version of our paper on how social &amp; behavioural science might support #COVID19 #coronavirus pandemic.…   

  user_screen_name  user_followers_count            user_location  \
0   snoopdogtanian                  2192                      NaN   
1        MAHandley                   335                      NaN   
2     steverathje2                  1204  University of Cambridge   

              user_name user_screen_name.1  user_time_zone  
0  Snoopdogtanian #UBI🌹     snoopdogtanian             NaN  
1      Margaret Handley          MAHandley             NaN  
2          Steve Rathje       steverathje2             NaN

## Explore

- number/% of geolocated tweets
- number/% of english language tweets (other languages to be dropped)
- number/% of uncommented retweets (to be dropped as simply duplicates)

In [16]:
# general look
tweets_raw.nunique()

created_at              12124
hashtags                 1630
favorite_count            118
id                      12161
reweet_id                1177
retweet_screen_name       861
lang                       22
place                      92
possibly_sensitive          2
retweet_count              90
text                     4480
user_screen_name         9381
user_followers_count     3994
user_location            3501
user_name                9291
user_screen_name.1       9381
user_time_zone              0
dtype: int64

There are 4,480 unique texts.

And 12,124 unique tweets.

### Uncommented retweets

To be dropped as they are exact duplicates of other tweets already in the dataset.

How to identify them?

- If original tweets was longer than 140 characters: (1) the tweet is a retweet (`retweet_id` is not `NaN`), and (2) the tweet ends in an ellipsis.
    **Rationale**: for retweets that do not include a comment where the original tweet is longer than 140 characters, the Twitter API returns a truncated full_text.  
    
- If original tweets was not longer than 140 characters: (1) the tweet is a retweet, (2) it starts with RT, and (3) its text is the exact duplication of an existing tweet

In [17]:
uncommented_above140_retweets = tweets_raw[tweets_raw.text.str.endswith("\u2026")]

In [18]:
uncommented_above140_retweets.shape

(8575, 17)

In [19]:
# how many also have a null reweet_id? Should be none
uncommented_above140_retweets.isnull().reweet_id.value_counts()    

False    8562
True       13
Name: reweet_id, dtype: int64

In [20]:
# let's see who these 13 are
uncommented_above140_retweets[uncommented_above140_retweets.isnull().reweet_id]

created_at  \
1871   Tue May 05 16:45:17 +0000 2020   
2349   Fri May 22 00:45:08 +0000 2020   
2529   Thu May 14 09:15:27 +0000 2020   
3709   Fri May 08 21:45:33 +0000 2020   
5088   Sat Apr 25 08:30:06 +0000 2020   
6007   Mon Mar 09 07:10:15 +0000 2020   
6228   Fri May 08 10:45:08 +0000 2020   
7256   Mon May 04 16:45:27 +0000 2020   
8040   Sat May 09 19:45:43 +0000 2020   
8128   Wed Mar 04 08:24:17 +0000 2020   
9257   Mon Apr 27 09:07:32 +0000 2020   
9480   Mon May 04 23:45:16 +0000 2020   
10130  Wed Mar 04 08:24:20 +0000 2020   

                                                                              hashtags  \
1871                             BehavioralScience education BehavioralScience COVID19   
2349   BehavioralScience education behaviouralscience behavioralscience socmar covid19   
2529                                                                   Covid19 covid19   
3709                                               BehavioralScience education COVID19   
5088                                                                           COVID19   
6007                                                               COVID19 Coronavirus   
6228                                               BehavioralScience education COVID19   
7256                             BehavioralScience education BehavioralScience COVID19   
8040                                               BehavioralScience education COVID19   
8128                                                               COVID19 Coronavirus   
9257                                              training behaviourchange Coronavirus   
9480                             BehavioralScience education BehavioralScience COVID19   
10130                                                                          COVID19   

       favorite_count                   id  reweet_id retweet_screen_name  \
1871                0  1257712986577350659        NaN                 NaN   
2349                0  1263631948536320006        NaN                 NaN   
2529                0  1260861271781650432        NaN                 NaN   
3709                0  1258875711609323522        NaN                 NaN   
5088                3  1253964490837811200        NaN                 NaN   
6007                0  1236912163677241344        NaN                 NaN   
6228                0  1258709513454129152        NaN                 NaN   
7256                0  1257350639916638210        NaN                 NaN   
8040                0  1259207943754526720        NaN                 NaN   
8128                0  1235118855648768000        NaN                 NaN   
9257                0  1254698684308209664        NaN                 NaN   
9480                1  1257456287551557632        NaN                 NaN   
10130               0  1235118869137600512        NaN                 NaN   

      lang place possibly_sensitive  retweet_count  \
1871    en   NaN                NaN              0   
2349    en   NaN                NaN              0   
2529    en   NaN              False              0   
3709    en   NaN                NaN              0   
5088    en   NaN              False              2   
6007    en   NaN                NaN              0   
6228    en   NaN                NaN              0   
7256    en   NaN                NaN              0   
8040    en   NaN                NaN              0   
8128    en   NaN                NaN              0   
9257    en   NaN              False              0   
9480    en   NaN                NaN              0   
10130   en   NaN                NaN              0   

                                                                                                                                                                                                                                                                                         text  \
1871                                                              

Note that many are variations of the same core message. They will probably be discounted as duplicates when we remove duplicates after pre-processing the text of the tweets.

In [21]:
# let's remove these 13 from the uncommented retweets as that's not what they are
uncommented_above140_retweets = uncommented_above140_retweets[uncommented_above140_retweets.notnull().reweet_id]

In [22]:
uncommented_above140_retweets.shape

(8562, 17)

In [23]:
tweets_to_go_ids1 = uncommented_above140_retweets.id

**Important NOTE**

The original tweets of these uncommented >140cha retweets may not be present in our dataset as original tweets. However, as for Twitter's policy, uncommented retweets of tweets longer than 140 character are truncated when accessed via the Stream API, so there is no way we can get access to the full text, we will have to exclude them anyway. We cannot, in fact, analyse the sentiment of incomplete text in a reliable way. 

Let's try to identify any other (i.e., up to 140 character) uncommented retweet

In [24]:
retweets = tweets_raw[(tweets_raw.text.str.startswith("RT @"))]

In [25]:
retweets.reweet_id.isnull().value_counts()

False    8829
True        3
Name: reweet_id, dtype: int64

In [26]:
# How many of these retweets are not in the above-140 uncommented rewteets already identified?
retweets[~retweets.id.isin(uncommented_above140_retweets.id)].shape


(270, 17)

In [27]:
# We need to understand which ones of these retweets are also uncommented dulicates of exisiting tweets (so to be dropped)
extra_retweets = retweets[~retweets.id.isin(uncommented_above140_retweets.id)]

In [28]:
# Let's clean the text by removing the "RT @name-original-sender " initial part
extra_retweets = extra_retweets.copy()
extra_retweets['cleaned_retweet'] = [re.sub(r"^RT @.*: ", "", t) for t in extra_retweets.text]

In [29]:
extra_retweets[['text', 'cleaned_retweet']]

text  \
35                                                                                                                                                                            RT @shayonislynn: Using #behaviouralscience to improve behaviours #nudgesinthewild #COVID19 https://t.co/OyIDElZwR7   
70                                                                                                                                                                   RT @faisalislam: Mark interviews the nudge unit behavioural scientist whose advice has driven the epidemic response so far -   
106                                                                                                                                                                  RT @faisalislam: Mark interviews the nudge unit behavioural scientist whose advice has driven the epidemic response so far -   
135    RT @TheLancet: NEW—Three lessons for #COVID19 response from pandemic #HIV \n\n1. Anticipate health inequalities\n2. Support behaviour change\n3. Multidisciplinary effort\n\nComment @TheLancetHIV by J Hargreaves, C Davey, et al @LSHTM\nhttps://t.co/8mPuOHL2tu https://t.co/zYROZ0epQE   
170                                                                                                                                                         RT @MDRC_News: Behavioural science and the response to #COVID19: A missed opportunity?\n\nvia @LSEImpactBlog  https://t.co/6Ybq1Wgb7N   
...                                                                                                                                                                                                                                                                                           ...   
12002                                                                                                                                                        RT @Christi55143759: Behaviour change and mass testing in areas with large nos who have tested positive will help contain the virus.   
12052                                                                                                                                                                RT @faisalislam: Mark interviews the nudge unit behavioural scientist whose advice has driven the epidemic response so far -   
12101                                                                                                                                                  RT @socratext: How social &amp; behavioural science might support the mitigation of #COVID19 #coronavirus pandemic @UKDCP @alexanderhaslam   
12142                                                                                                                                                                RT @faisalislam: Mark interviews the nudge unit behavioural scientist whose advice has driven the epidemic response so far -   
12160                                                                                                                                                       RT @EUnotCoup: @jclanoe @KnittedKittie Medley &amp; Halpern (head of the Nudge Unit) are big on #HerdImmunity https://t.co/vF6m4iQTp4   

                                                                                                                                                                                                                                                                   cleaned_retweet  
35                                                                                                                                                                               Using #behaviouralscience to improve behaviours #nudgesinthewild #COVID19 https://t.co/OyIDElZwR7  
70                                                                                                                                                                     Mark interviews the nudge unit behavioural scie

In [30]:
pd.DataFrame(extra_retweets.cleaned_retweet.value_counts())

cleaned_retweet
Mark interviews the nudge unit behavioural scientist whose advice has driven the epidemic response so far -                           97
Forget all behavioural science stuff. Look at the number of people on respirators - that’s the life saving element                    35
Paging Nudge Unit:\nTry bundling your infection control messages.\n\nhttps://t.co/2pJeBihyge                                          16
drivers plan to walk for cleaner air #greenrecovery https://t.co/6K3C8cvYFP                                                            6
Thank god for real experts on Pandemics and Healthcare, rather Nudge Unit behavioural scientists.                                      4
...                                                                                                                                  ...
And another member of the behavioural science advisory group speaks out 👇\n\n#COVID19 #pandemic https://t.co/j0ENdOSlJq                1
Calculate your #covid-19 risk and get some behaviour change advice too https://t.co/7zhQ4M91Cg https://t.co/id8V5Jy2e8                 1
This looks like the triumph of 'nudge theory' over medical science.                                                                    1
Using #behaviouralscience to improve behaviours #nudgesinthewild #COVID19 https://t.co/OyIDElZwR7                                      1
What have you found is the easiest behaviour change to implement?\n\n#energysavingtips #COVID19 #stayhome                              1

[73 rows x 1 columns]

Some, those whose count > 1, are definitely uncommented retweets of indentical original tweets - to be removed.

In [38]:
# let's keep one of each, in case the original tweet is not present in the dataset
extra_retweets.drop_duplicates(subset ="cleaned_retweet", keep = "first")

created_at  \
71     Sun Mar 08 17:00:22 +0000 2020   
95     Tue Apr 07 20:37:48 +0000 2020   
177    Sat Mar 14 19:22:54 +0000 2020   
227    Thu Mar 12 22:52:22 +0000 2020   
403    Sat Apr 18 12:50:57 +0000 2020   
...                               ...   
13964  Fri Apr 17 22:28:46 +0000 2020   
14090  Fri Apr 24 09:55:03 +0000 2020   
14350  Tue Apr 28 11:02:20 +0000 2020   
14505  Thu Apr 09 19:35:53 +0000 2020   
14554  Fri Apr 10 08:55:42 +0000 2020   

                                           hashtags  favorite_count  \
71                                              NaN               0   
95         transportation BehavioralScience COVID19               0   
177                                             NaN               0   
227                                             NaN               0   
403                        uspol regulation COVID19               0   
...                                             ...             ...   
13964                                       covid19               0   
14090                   COVID19 BehavioralEconomics               3   
14350  stayathome coronavirus behaviourchange newmr               0   
14505                           COVID19 coronavirus               0   
14554                Behaviouralinsight coronavirus               0   

                        id     reweet_id retweet_screen_name lang place  \
71     1236698282736472065  1.236694e+18           normonics   fr   NaN   
95     1247624638588387331  1.247618e+18         brianne_eby   en   NaN   
177    1238908479718465538  1.238906e+18          peterjukes   en   NaN   
227    1238236419124088833  1.238092e+18         faisalislam   en   NaN   
403    1251493421367218178  1.251492e+18            DrRimmer   en   NaN   
...                    ...           ...                 ...  ...   ...   
13964  1251276443469770753  1.251105e+18          JulieLeask   en   NaN   
14090  1253623480718970880           NaN                 NaN   en   NaN   
14350  1255089963521019904  1.255089e+18           neilgains   en   NaN   
14505  1248333832937517062  1.248245e+18           socratext   en   NaN   
14554  1248535113094369280  1.248535e+18        TheDanWilson   en   NaN   

      possibly_sensitive  retweet_count  \
71                 False              2   
95                 False              4   
177                  NaN             90   
227                  NaN            191   
403                False              3   
...                  ...            ...   
13964                NaN              3   
14090              False              1   
14350              False              1   
14505                NaN              2   
14554              False              1   

                                                                                                                                                                                                                                                                                              text  \
71                                                                                                                                                                                                         RT @normonics: .@casssunstein nudge in action. Hyper dangerous. https://t.co/iYnfuV1ED1   
95                                                                                                                                               RT @brianne_eby: I wrote about the nexus between #transportation &amp; #BehavioralScience in the context of #COVID19 👇\n\nhttps://t.co/aVDKN3UY3W   
177                                                                                                                                                             RT @peterjukes: Forget all behavioural science stuff. Look at the number of people on respirators - that’s the life saving element   
227                                                                                    

In [31]:
# ids of duplicate retweets (to be removed)
tweets_to_go_ids2 = extra_retweets[extra_retweets.duplicated('cleaned_retweet')].id

In [32]:
tweets_to_go_ids2

106      1245224046196690945
328      1238100680587579394
574      1238913289058484225
688      1238100459505758210
993      1266325354911801345
                ...         
11866    1264813914208026626
11868    1247085494103674881
11932    1238400099795042304
12052    1238400129142579201
12142    1238128929422131201
Name: id, Length: 197, dtype: int64

In [33]:
# What's left?
extra_retweets[~extra_retweets.id.isin(tweets_to_go_ids2)]

created_at  \
35     Thu Mar 19 19:58:53 +0000 2020   
70     Thu Mar 12 16:35:45 +0000 2020   
135    Fri Apr 17 14:21:45 +0000 2020   
170    Sat May 30 16:45:55 +0000 2020   
277    Sat Apr 04 06:52:45 +0000 2020   
...                               ...   
11561  Sun Apr 05 09:09:33 +0000 2020   
11943  Tue Apr 28 11:02:20 +0000 2020   
12002  Wed May 06 18:08:30 +0000 2020   
12101  Thu Apr 09 19:35:53 +0000 2020   
12160  Fri May 29 23:06:40 +0000 2020   

                                           hashtags  favorite_count  \
35       behaviouralscience nudgesinthewild COVID19               0   
70                                              NaN               0   
135                                     COVID19 HIV               1   
170                                         COVID19               0   
277                                         COVID19               0   
...                                             ...             ...   
11561                 BigData Google COVID19 Anroid               0   
11943  stayathome coronavirus behaviourchange newmr               0   
12002                                           NaN               0   
12101                           COVID19 coronavirus               0   
12160                                  HerdImmunity               0   

                        id     reweet_id retweet_screen_name lang place  \
35     1240729478088790016  1.240444e+18        shayonislynn   en   NaN   
70     1238141642005254148  1.238092e+18         faisalislam   en   NaN   
135    1251153881960046595           NaN                 NaN   en   NaN   
170    1266772841967620096  1.266325e+18           MDRC_News   en   NaN   
277    1246329847472848897  1.246162e+18           bgibson49   en   NaN   
...                    ...           ...                 ...  ...   ...   
11561  1246726661866848256  1.246719e+18          mirko_ross   en   NaN   
11943  1255089963521019904  1.255089e+18           neilgains   en   NaN   
12002  1258096316116279296  1.257669e+18     Christi55143759   en   NaN   
12101  1248333832937517062  1.248245e+18           socratext   en   NaN   
12160  1266506273031421952  1.266506e+18           EUnotCoup   en   NaN   

      possibly_sensitive  retweet_count  \
35                 False              4   
70                   NaN            189   
135                False              0   
170                False              2   
277                  NaN              1   
...                  ...            ...   
11561              False              1   
11943              False              1   
12002                NaN              1   
12101                NaN              2   
12160              False              1   

                                                                                                                                                                                                                                                                                             text  \
35                                                                                                                                                                            RT @shayonislynn: Using #behaviouralscience to improve behaviours #nudgesinthewild #COVID19 https://t.co/OyIDElZwR7   
70                                                                                                                                                                   RT @faisalislam: Mark interviews the nudge unit behavioural scientist whose advice has driven the epidemic response so far -   
135    RT @TheLancet: NEW—Three lessons for #COVID19 response from pandemic #HIV \n\n1. Anticipate health inequalities\n2. Support behaviour change\n3. Multidisciplinary effort\n\nComment @TheLancetHIV by J Hargreaves, C Davey, et al @LSHTM\nhttps://t.co/8mPuOHL2tu https://t.co/zYROZ0epQE   
170                                                                                        

In [34]:
# are they maybe duplicates of tweets of original tweets in the dataset?
tweets_to_investigate = extra_retweets[~extra_retweets.id.isin(tweets_to_go_ids2)]

In [35]:
tweets_to_investigate_text = tweets_to_investigate.cleaned_retweet     

In [36]:
len(tweets_to_investigate_text)

73

In [37]:
tweets_raw[tweets_raw.text.isin(tweets_to_investigate_text)].count()   # 35 are "copies" of original tweets in the dataset

created_at              26
hashtags                20
favorite_count          26
id                      26
reweet_id                0
retweet_screen_name      0
lang                    26
place                    0
possibly_sensitive      23
retweet_count           26
text                    26
user_screen_name        26
user_followers_count    26
user_location           23
user_name               26
user_screen_name.1      26
user_time_zone           0
dtype: int64

In [38]:
pd.DataFrame(tweets_to_investigate_text.value_counts())

cleaned_retweet
Paging Nudge Unit:\nTry bundling your infection control messages.\n\nhttps://t.co/2pJeBihyge                                               1
Forget all behavioural science stuff. Look at the number of people on respirators - that’s the life saving element                         1
It will be very important to get good evidence on this @PHE_uk @NCSCT @UCL_BSH @SusanMichie                                                1
👀 #BigData by #Google on #COVID19 behaviour change of #Anroid users...\n\nhttps://t.co/ohgOyt92a5                                          1
Such a pleasure to be part of this fantastic group of authors. Now in press at Nature Behavioural Science.                                 1
...                                                                                                                                      ...
How do #behaviouralscience and testing aid responses to #coronavirus ?  https://t.co/FnMnT0NFNO https://t.co/QSvahdf0vu                    1
How social &amp; behavioural science might support the mitigation of #COVID19 #coronavirus pandemic @UKDCP @alexanderhaslam                1
#SocialDistance on #QandA tonight.  Appropriate behaviour change for #COVID19 https://t.co/Fkz9CWOA0v                                      1
a missed opportunity?\n\n#COVID19 https://t.co/vvijU3DEU2                                                                                  1
HE IS NOT CORONAVIRUS SCIENCE ADVISOR HE IS A PSYCHOLOGIST AND HEAD OF THE FUCKING NUDGE UNIT WHAT THE FUCK ARE YOU DOING                  1

[73 rows x 1 columns]

In [39]:
tweets_to_go_ids3 = tweets_to_investigate[tweets_to_investigate.cleaned_retweet.isin(
    tweets_raw[tweets_raw.text.isin(tweets_to_investigate_text)].text)].id

In [40]:
len(tweets_to_go_ids3)   #ok

26

In [41]:
print(len(tweets_to_go_ids1))
print(len(tweets_to_go_ids2))

8562
197


### Let's remove from the dataset all the uncommented retweets that are duplicate of original tweets already in the dataset 

In [42]:
print(len(set(tweets_to_go_ids1)))
print(len(set(tweets_to_go_ids2)))
print(len(set(tweets_to_go_ids3)))
# GOOD :-), each is unique

8562
197
26


In [43]:
tweets_to_go_ids_all = list(tweets_to_go_ids1) + list(tweets_to_go_ids2) + list(tweets_to_go_ids3)

In [44]:
len(tweets_to_go_ids_all)

8785

In [45]:
len(tweets_raw.index) 

12161

In [46]:
tweets_original = tweets_raw[~tweets_raw.id.isin(tweets_to_go_ids_all)]

In [47]:
len(tweets_original.index)

3376

### English vs. non-English tweets

Non-English tweets will be dropped as they are not part of our population of interest.


In [48]:
count_by_lang = pd.DataFrame(tweets_original.groupby('lang').id.count()) 

In [49]:
count_by_lang.rename(columns={'id': 'count_ids'}, inplace=True)

In [50]:
count_by_lang['prop'] = round(count_by_lang['count_ids'] / sum(count_by_lang['count_ids']),3)

In [51]:
count_by_lang

count_ids   prop
lang                  
ar            1  0.000
cy            1  0.000
da            2  0.001
de            5  0.001
en         3107  0.920
es           54  0.016
fa            1  0.000
fi            2  0.001
fr            9  0.003
hi            3  0.001
ht            1  0.000
in            1  0.000
it           10  0.003
nl            4  0.001
no            1  0.000
pl            2  0.001
pt           10  0.003
ru            1  0.000
sv            3  0.001
tl            1  0.000
tr            5  0.001
und         152  0.045

What are the "und" (undefined) ones?

In [52]:
tweets_original[tweets_original.lang == "und"]

created_at  \
99     Tue Apr 21 22:09:59 +0000 2020   
236    Fri Mar 13 12:11:49 +0000 2020   
339    Wed May 27 16:15:56 +0000 2020   
449    Sun Jun 07 10:43:06 +0000 2020   
454    Mon Mar 09 22:17:05 +0000 2020   
...                               ...   
11926  Mon May 04 15:51:34 +0000 2020   
11953  Thu Apr 02 08:11:22 +0000 2020   
12056  Fri Mar 13 09:42:52 +0000 2020   
12092  Fri Mar 13 09:42:54 +0000 2020   
12156  Fri Apr 10 06:32:47 +0000 2020   

                                                                                                                                         hashtags  \
99                                                                                                                                            NaN   
236    Coronavirus BehaviouralScience MoneyBeforeLives CullTheVulnerable TorySocialCleransing ToryEugenics HelpUs MurderingToryScum BorisOut GTTO   
339                                                                                                                                           NaN   
449                                                                            marr MarrShow Dispatches cummingstogetya HerdImmunity dreddjustice   
454                                                                                                                                           NaN   
...                                                                                                                                           ...   
11926                                                                                                                                 ApnoKaDhyan   
11953                                                                            BorisResign BorisOut HerdimmunityIsMurder ukcoronavirus COVIDー19   
12056                                                                                                                                         NaN   
12092                                                                                                                  CoronaOutbreak coronavirus   
12156                                                                                                                                         NaN   

       favorite_count                   id  reweet_id retweet_screen_name  \
99                  0  1252721267339190272        NaN                 NaN   
236                 3  1238437609749258241        NaN                 NaN   
339                 0  1265678130896867328        NaN                 NaN   
449                 0  1269580639105355776        NaN                 NaN   
454                 2  1237140377569759233        NaN                 NaN   
...               ...                  ...        ...                 ...   
11926               5  1257337079203315719        NaN                 NaN   
11953               0  1245624852179902469        NaN                 NaN   
12056               2  1238400124394627073        NaN                 NaN   
12092               0  1238400131516575744        NaN                 NaN   
12156               0  1248499146970955776        NaN                 NaN   

      lang place possibly_sensitive  retweet_count  \
99     und   NaN              False              0   
236    und   NaN              False              2   
339    und   NaN              False              0   
449    und   NaN               True              0   
454    und   NaN              False              0   
...    ...   ...                ...            ...   
11926  und   NaN              False              0   
11953  und   NaN              False              0   
12056  und   NaN              False              0   
12092  und   NaN              False              0   
12156  und   NaN              False              0   

                                                                                                                                                                           

They seem to be purely list of hashtags and duplicates of the same one. So let's exclude them from further investigation.

### Exclude all non-English tweets

In [53]:
tweets_original_en = tweets_original[tweets_original.lang == 'en']

In [54]:
len(tweets_original_en.index)

3107

### Geolocation

We'll look at the geolocation of users.

In [55]:
tweets_original_en.place.value_counts(dropna=False)

NaN                     2982
Geelong, Victoria          9
Dunstable, England         7
Blackpool, England         6
North East, England        3
                        ... 
Croydon, London            1
Chelsea, Québec            1
Halifax, Nova Scotia       1
Bampton, England           1
Richmond, London           1
Name: place, Length: 87, dtype: int64

The large majority are not geolocated.

### Save the data

In [56]:
# define file name and file path
output_name = "tweets_original_en"
output_filepath = os.path.join(OUTPUT_DIR, output_name + ".csv")

In [57]:
tweets_original_en.to_csv(output_filepath)